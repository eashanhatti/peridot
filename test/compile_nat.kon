let {
  datatype Nat : OType {
    zero : Nat;
    succ : '[_ : Nat] -> Nat;
  };

  assume CompileNat : [_ : (LiftL Int)] -> [_ : (LiftC Nat)] -> MType;
  assume compile_zero : <CompileNat (quoteL 0) (quoteC zero)>;
  assume compile_succ :
    [x : (LiftL Int)] -> [n : (LiftC Nat)] ->
    rule <CompileNat (quoteL (+ (spliceL x) 2)) (quoteC <succ (spliceC n)>)> :-
      <CompileNat x n>;

  fresh x : (LiftL Int);
  prove <CompileNat <succ <succ zero>> x>;
} in {
  OType
}
