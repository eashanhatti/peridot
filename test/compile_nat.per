define Nat
  : Type
  = Record {
      tag: Bool,
      data:
        if tag {
          Record {}
        } else {
          Nat
        }
    };

define zero
  : Nat
  = record {
      tag = true,
      data = record {}
    };

define succ
  : Function(Nat) -> Nat
  = function(n) =>
      record {
        tag = false,
        data = n
      };

axiom CompileNat: Function(CCode(C_Int), Code(Nat)) ~> MetaType;
axiom compile_zero : CompileNat(c<[c_int 0]>, <zero>);
axiom compile_succ:
  Forall x, Forall n,
  Implies(
    CompileNat(x, n),
    CompileNat(c<[c_add c~x [c_int 1]]>, <succ(~n)>));

variable n: Code(Nat);
proof CompileNat(c<[c_add [c_add [c_int 0] [c_int 1]] [c_int 1]]>, <succ(~n)>);
