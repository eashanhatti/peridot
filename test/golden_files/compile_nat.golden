((fromList
    [(UVGlobal 6,
      Just
        (Rigid
           (CodeObjIntro
              (Neutral
                 _
                 (ObjFunElim
                    (Neutral
                       _
                       (GlobalVar (Rigid (NameIntro Obj (Id {unId = 2})))))
                    (Neutral
                       _
                       (CodeObjElim
                          (Neutral _ (UniVar (LVGlobal 2001)))))))))),
     (UVGlobal 7,
      Just
        (Rigid
           (CodeObjIntro
              (Neutral
                 _
                 (GlobalVar
                    (Rigid (NameIntro Obj (Id {unId = 1})))))))),
     (UVGlobal 1000, Just (Rigid (CIntType))),
     (UVGlobal 1001,
      Just
        (Neutral
           _
           (GlobalVar
              (Rigid (NameIntro Obj (Id {unId = 0})))))),
     (UVGlobal 1002, Nothing),
     (UVGlobal 1003,
      Just (Rigid (CodeCType (Rigid (CIntType))))),
     (UVGlobal 1004, Just (Rigid (TypeType Meta))),
     (UVGlobal 1005, Nothing),
     (UVGlobal 1006,
      Just
        (Rigid
           (CodeObjType
              (Neutral
                 _
                 (GlobalVar
                    (Rigid (NameIntro Obj (Id {unId = 0})))))))),
     (UVGlobal 1007, Just (Rigid (TypeType Meta))),
     (UVGlobal 1008, Nothing),
     (UVGlobal 1009,
      Just (Rigid (CodeCType (Rigid (CIntType))))),
     (UVGlobal 1010, Just (Rigid (TypeType Meta))),
     (UVGlobal 1011, Nothing),
     (UVGlobal 1012,
      Just
        (Rigid
           (CodeObjType
              (Neutral
                 _
                 (GlobalVar
                    (Rigid (NameIntro Obj (Id {unId = 0})))))))),
     (UVGlobal 1013, Just (Rigid (TypeType Meta))),
     (UVGlobal 1014, Just (Rigid (CIntType))),
     (UVGlobal 1015,
      Just
        (Neutral
           _
           (GlobalVar
              (Rigid (NameIntro Obj (Id {unId = 0})))))),
     (UVGlobal 1016, Nothing),
     (UVGlobal 1017,
      Just (Rigid (CodeCType (Rigid (CIntType))))),
     (UVGlobal 1018, Just (Rigid (TypeType Meta))),
     (UVGlobal 1019, Nothing),
     (UVGlobal 1020,
      Just
        (Rigid
           (CodeObjType
              (Neutral
                 _
                 (GlobalVar
                    (Rigid (NameIntro Obj (Id {unId = 0})))))))),
     (UVGlobal 1021, Just (Rigid (TypeType Meta))),
     (UVGlobal 1022, Just (Rigid (CIntType))),
     (UVGlobal 1023,
      Just
        (Neutral
           _
           (GlobalVar
              (Rigid (NameIntro Obj (Id {unId = 0})))))),
     (UVGlobal 1024, Nothing),
     (UVGlobal 1025,
      Just (Rigid (CodeCType (Rigid (CIntType))))),
     (UVGlobal 1026, Just (Rigid (TypeType Meta))),
     (UVGlobal 1027, Nothing),
     (UVGlobal 1028,
      Just
        (Rigid
           (CodeObjType
              (Neutral
                 _
                 (GlobalVar
                    (Rigid (NameIntro Obj (Id {unId = 0})))))))),
     (UVGlobal 1029, Just (Rigid (TypeType Meta))),
     (UVGlobal 1030, Just (Rigid (CIntType))),
     (UVGlobal 1031,
      Just
        (Neutral
           _
           (GlobalVar
              (Rigid (NameIntro Obj (Id {unId = 0})))))),
     (UVGlobal 1032, Just (Rigid (CIntType))),
     (UVGlobal 1033,
      Just
        (Neutral
           _
           (GlobalVar
              (Rigid (NameIntro Obj (Id {unId = 0})))))),
     (UVGlobal 1034, Nothing),
     (UVGlobal 1035,
      Just (Rigid (CodeCType (Rigid (CIntType))))),
     (UVGlobal 1036, Just (Rigid (TypeType Meta))),
     (UVGlobal 1037, Nothing),
     (UVGlobal 1038,
      Just
        (Rigid
           (CodeObjType
              (Neutral
                 _
                 (GlobalVar
                    (Rigid (NameIntro Obj (Id {unId = 0})))))))),
     (UVGlobal 1039, Just (Rigid (TypeType Meta))),
     (LVGlobal 2000,
      Just (Rigid (CodeCIntro (Rigid (CIntIntro 0))))),
     (LVGlobal 2001,
      Just
        (Neutral
           _
           (GlobalVar
              (Rigid (NameIntro Meta (Id {unId = 7}))))))],
  fromList
    [(UVGlobal 1002, UVGlobal 1003),
     (UVGlobal 1003, UVGlobal 1002),
     (UVGlobal 1005, UVGlobal 1006),
     (UVGlobal 1006, UVGlobal 1005),
     (UVGlobal 1008, UVGlobal 1009),
     (UVGlobal 1009, UVGlobal 1008),
     (UVGlobal 1011, UVGlobal 1012),
     (UVGlobal 1012, UVGlobal 1011),
     (UVGlobal 1016, UVGlobal 1017),
     (UVGlobal 1017, UVGlobal 1016),
     (UVGlobal 1019, UVGlobal 1020),
     (UVGlobal 1020, UVGlobal 1019),
     (UVGlobal 1024, UVGlobal 1025),
     (UVGlobal 1025, UVGlobal 1024),
     (UVGlobal 1027, UVGlobal 1028),
     (UVGlobal 1028, UVGlobal 1027),
     (UVGlobal 1034, UVGlobal 1035),
     (UVGlobal 1035, UVGlobal 1034),
     (UVGlobal 1037, UVGlobal 1038),
     (UVGlobal 1038, UVGlobal 1037)],
  fromList
    [(SourcePos
        {sourceName = "<TODO>",
         sourceLine = Pos 42,
         sourceColumn = Pos 1},
      FailedProve
        (Neutral
           _
           (MetaFunElim
              (Neutral
                 _
                 (MetaFunElim
                    (Neutral
                       _
                       (GlobalVar (Rigid (NameIntro Meta (Id {unId = 3})))))
                    (Rigid
                       (CodeCIntro
                          (Rigid
                             (CIntElimAdd
                                (Rigid (CIntIntro 0))
                                (Rigid (CIntIntro 1))))))))
              (Rigid
                 (CodeObjIntro
                    (Neutral
                       _
                       (ObjFunElim
                          (Neutral
                             _
                             (GlobalVar
                                (Rigid (NameIntro Obj (Id {unId = 2})))))
                          (Neutral
                             _
                             (ObjFunElim
                                (Neutral
                                   _
                                   (GlobalVar
                                      (Rigid (NameIntro Obj (Id {unId = 2})))))
                                (Neutral
                                   _
                                   (CodeObjElim
                                      (Neutral
                                         _
                                         (GlobalVar
                                            (Rigid
                                               (NameIntro
                                                  Meta
                                                  (Id
                                                     {unId =
                                                        8}))))))))))))))))]),
 Declare
   Obj
   (Rigid (NameIntro Obj (Id {unId = 0})))
   (Rigid (TypeType Obj))
   (Define
      (Rigid (NameIntro Obj (Id {unId = 0})))
      (RecType
         (fromList
            [(Field {unField = "tag"}, Rigid TwoType),
             (Field {unField = "data"},
              TwoElim
                (LocalVar (Index {unIndex = 0}))
                (RecType (fromList []))
                (GlobalVar
                   (Rigid (NameIntro Obj (Id {unId = 0})))))]))
      (Declare
         Obj
         (Rigid (NameIntro Obj (Id {unId = 1})))
         (GlobalVar (Rigid (NameIntro Obj (Id {unId = 0}))))
         (Declare
            Obj
            (Rigid (NameIntro Obj (Id {unId = 2})))
            (ObjFunType
               Explicit
               (GlobalVar (Rigid (NameIntro Obj (Id {unId = 0}))))
               (GlobalVar (Rigid (NameIntro Obj (Id {unId = 0})))))
            (Declare
               Meta
               (Rigid (NameIntro Meta (Id {unId = 3})))
               (MetaFunType
                  Explicit
                  (Rigid (CodeCType (Rigid CIntType)))
                  (MetaFunType
                     Explicit
                     (Rigid
                        (CodeObjType
                           (GlobalVar (Rigid (NameIntro Obj (Id {unId = 0}))))))
                     (Rigid (TypeType Meta))))
               (Declare
                  Meta
                  (Rigid (NameIntro Meta (Id {unId = 6})))
                  (Rigid
                     (CodeObjType
                        (GlobalVar (Rigid (NameIntro Obj (Id {unId = 0}))))))
                  (Declare
                     Meta
                     (Rigid (NameIntro Meta (Id {unId = 7})))
                     (Rigid
                        (CodeObjType
                           (GlobalVar (Rigid (NameIntro Obj (Id {unId = 0}))))))
                     (Declare
                        Meta
                        (Rigid (NameIntro Meta (Id {unId = 8})))
                        (Rigid
                           (CodeObjType
                              (GlobalVar
                                 (Rigid (NameIntro Obj (Id {unId = 0}))))))
                        (Define
                           (Rigid (NameIntro Obj (Id {unId = 1})))
                           (RecIntro
                              (fromList
                                 [(Field {unField = "tag"}, Rigid TwoIntro0),
                                  (Field {unField = "data"},
                                   RecIntro (fromList []))]))
                           (Define
                              (Rigid (NameIntro Obj (Id {unId = 2})))
                              (ObjFunIntro
                                 (RecIntro
                                    (fromList
                                       [(Field {unField = "tag"},
                                         Rigid TwoIntro1),
                                        (Field {unField = "data"},
                                         LocalVar (Index {unIndex = 1}))])))
                              (Define
                                 (Rigid (NameIntro Meta (Id {unId = 3})))
                                 (Rigid (MetaConstIntro (Id {unId = 3})))
                                 (Define
                                    (Rigid (NameIntro Meta (Id {unId = 6})))
                                    (UniVar (UVGlobal 6))
                                    (Define
                                       (Rigid (NameIntro Meta (Id {unId = 7})))
                                       (UniVar (UVGlobal 7))
                                       (Define
                                          (Rigid
                                             (NameIntro Meta (Id {unId = 8})))
                                          (UniVar (UVGlobal 8))
                                          (Declare
                                             Meta
                                             (Rigid
                                                (NameIntro
                                                   Meta (Id {unId = 4})))
                                             (MetaFunElim
                                                (MetaFunElim
                                                   (GlobalVar
                                                      (Rigid
                                                         (NameIntro
                                                            Meta
                                                            (Id {unId = 3}))))
                                                   (Rigid
                                                      (CodeCIntro
                                                         (Rigid
                                                            (CIntIntro 0)))))
                                                (Rigid
                                                   (CodeObjIntro
                                                      (GlobalVar
                                                         (Rigid
                                                            (NameIntro
                                                               Obj
                                                               (Id
                                                                  {unId =
                                                                     1})))))))
                                             (Declare
                                                Meta
                                                (Rigid
                                                   (NameIntro
                                                      Meta (Id {unId = 5})))
                                                (Rigid
                                                   (AllType
                                                      (MetaFunIntro
                                                         (Rigid
                                                            (AllType
                                                               (MetaFunIntro
                                                                  (Rigid
                                                                     (ImplType
                                                                        (MetaFunElim
                                                                           (MetaFunElim
                                                                              (GlobalVar
                                                                                 (Rigid
                                                                                    (NameIntro
                                                                                       Meta
                                                                                       (Id
                                                                                          {unId =
                                                                                             3}))))
                                                                              (LocalVar
                                                                                 (Index
                                                                                    {unIndex =
                                                                                       1})))
                                                                           (LocalVar
                                                                              (Index
                                                                                 {unIndex =
                                                                                    0})))
                                                                        (MetaFunElim
                                                                           (MetaFunElim
                                                                              (GlobalVar
                                                                                 (Rigid
                                                                                    (NameIntro
                                                                                       Meta
                                                                                       (Id
                                                                                          {unId =
                                                                                             3}))))
                                                                              (Rigid
                                                                                 (CodeCIntro
                                                                                    (Rigid
                                                                                       (CIntElimAdd
                                                                                          (CodeCElim
                                                                                             (LocalVar
                                                                                                (Index
                                                                                                   {unIndex =
                                                                                                      1})))
                                                                                          (Rigid
                                                                                             (CIntIntro
                                                                                                1)))))))
                                                                           (Rigid
                                                                              (CodeObjIntro
                                                                                 (ObjFunElim
                                                                                    (GlobalVar
                                                                                       (Rigid
                                                                                          (NameIntro
                                                                                             Obj
                                                                                             (Id
                                                                                                {unId =
                                                                                                   2}))))
                                                                                    (CodeObjElim
                                                                                       (LocalVar
                                                                                          (Index
                                                                                             {unIndex =
                                                                                                0})))))))))))))))
                                                (Declare
                                                   Meta
                                                   (Rigid
                                                      (NameIntro
                                                         Meta (Id {unId = 9})))
                                                   (MetaFunElim
                                                      (MetaFunElim
                                                         (GlobalVar
                                                            (Rigid
                                                               (NameIntro
                                                                  Meta
                                                                  (Id
                                                                     {unId =
                                                                        3}))))
                                                         (Rigid
                                                            (CodeCIntro
                                                               (Rigid
                                                                  (CIntElimAdd
                                                                     (Rigid
                                                                        (CIntIntro
                                                                           0))
                                                                     (Rigid
                                                                        (CIntIntro
                                                                           1)))))))
                                                      (GlobalVar
                                                         (Rigid
                                                            (NameIntro
                                                               Meta
                                                               (Id
                                                                  {unId =
                                                                     6})))))
                                                   (Declare
                                                      Meta
                                                      (Rigid
                                                         (NameIntro
                                                            Meta
                                                            (Id {unId = 10})))
                                                      (MetaFunElim
                                                         (MetaFunElim
                                                            (GlobalVar
                                                               (Rigid
                                                                  (NameIntro
                                                                     Meta
                                                                     (Id
                                                                        {unId =
                                                                           3}))))
                                                            (Rigid
                                                               (CodeCIntro
                                                                  (Rigid
                                                                     (CIntElimAdd
                                                                        (Rigid
                                                                           (CIntIntro
                                                                              0))
                                                                        (Rigid
                                                                           (CIntIntro
                                                                              1)))))))
                                                         (Rigid
                                                            (CodeObjIntro
                                                               (ObjFunElim
                                                                  (GlobalVar
                                                                     (Rigid
                                                                        (NameIntro
                                                                           Obj
                                                                           (Id
                                                                              {unId =
                                                                                 2}))))
                                                                  (CodeObjElim
                                                                     (GlobalVar
                                                                        (Rigid
                                                                           (NameIntro
                                                                              Meta
                                                                              (Id
                                                                                 {unId =
                                                                                    7})))))))))
                                                      (Declare
                                                         Meta
                                                         (Rigid
                                                            (NameIntro
                                                               Meta
                                                               (Id
                                                                  {unId = 11})))
                                                         (MetaFunElim
                                                            (MetaFunElim
                                                               (GlobalVar
                                                                  (Rigid
                                                                     (NameIntro
                                                                        Meta
                                                                        (Id
                                                                           {unId =
                                                                              3}))))
                                                               (Rigid
                                                                  (CodeCIntro
                                                                     (Rigid
                                                                        (CIntElimAdd
                                                                           (Rigid
                                                                              (CIntIntro
                                                                                 0))
                                                                           (Rigid
                                                                              (CIntIntro
                                                                                 1)))))))
                                                            (Rigid
                                                               (CodeObjIntro
                                                                  (ObjFunElim
                                                                     (GlobalVar
                                                                        (Rigid
                                                                           (NameIntro
                                                                              Obj
                                                                              (Id
                                                                                 {unId =
                                                                                    2}))))
                                                                     (ObjFunElim
                                                                        (GlobalVar
                                                                           (Rigid
                                                                              (NameIntro
                                                                                 Obj
                                                                                 (Id
                                                                                    {unId =
                                                                                       2}))))
                                                                        (CodeObjElim
                                                                           (GlobalVar
                                                                              (Rigid
                                                                                 (NameIntro
                                                                                    Meta
                                                                                    (Id
                                                                                       {unId =
                                                                                          8}))))))))))
                                                         (Define
                                                            (Rigid
                                                               (NameIntro
                                                                  Meta
                                                                  (Id
                                                                     {unId =
                                                                        4})))
                                                            (Rigid
                                                               (MetaConstIntro
                                                                  (Id
                                                                     {unId =
                                                                        4})))
                                                            (Define
                                                               (Rigid
                                                                  (NameIntro
                                                                     Meta
                                                                     (Id
                                                                        {unId =
                                                                           5})))
                                                               (Rigid
                                                                  (MetaConstIntro
                                                                     (Id
                                                                        {unId =
                                                                           5})))
                                                               (Define
                                                                  (Rigid
                                                                     (NameIntro
                                                                        Meta
                                                                        (Id
                                                                           {unId =
                                                                              9})))
                                                                  (Rigid
                                                                     (MetaConstIntro
                                                                        (Id
                                                                           {unId =
                                                                              9})))
                                                                  (Define
                                                                     (Rigid
                                                                        (NameIntro
                                                                           Meta
                                                                           (Id
                                                                              {unId =
                                                                                 10})))
                                                                     (Rigid
                                                                        (MetaConstIntro
                                                                           (Id
                                                                              {unId =
                                                                                 10})))
                                                                     (Define
                                                                        (Rigid
                                                                           (NameIntro
                                                                              Meta
                                                                              (Id
                                                                                 {unId =
                                                                                    11})))
                                                                        (Rigid
                                                                           ElabError)
                                                                        (Rigid
                                                                           (TypeType
                                                                              Obj))))))))))))))))))))))))))
